File: Go Files/sorting/insertionsort.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

func InsertionSort(arr []int) {
	leng := len(arr)

	for i := 1; i < leng; i++ {
		key := arr[i]
		j := i - 1

		for j >= 0 && arr[j] > key {
			arr[j+1] = arr[j]
			j--
		}

		arr[j+1] = key
	}
}


func generateRandomArray(size int, maxVal int) []int {

	rand.Seed(time.Now().UnixNano()) // Seed 
    
	arr := make([]int, size)

	for i := 0; i < size; i++ {
		arr[i] = rand.Intn(maxVal)
	}

	return arr
}


func main() {
	size := 100000       
	maxVal := 100     
	arr := generateRandomArray(size, maxVal)

	start := time.Now()
	InsertionSort(arr)
    end := time.Now()

    fmt.Println("Insertion Sort: ", end.Sub(start))
	
}

File: C-files/runtime speed/sorting/insertionsort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>


void insertionSort(int arr[], int n) {
    for (int i = 1; i < n; i++) {
        int key = arr[i];
        int j = i - 1;

        while (j >= 0 && arr[j] > key) { // Shift elements to the right
            arr[j + 1] = arr[j];
            j--;
        }

        arr[j + 1] = key;
    }
}


void generateRandomArray(int arr[], int n, int maxVal) {
    for (int i = 0; i < n; i++) {
        arr[i] = rand() % maxVal; // Random number between 0 and maxVal-1
    }
}


void printArray(int arr[], int n) {
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

int main() {
    int n = 100000;   
    int maxVal = 1000; 

    int *arr = (int *)malloc(n * sizeof(int)); 
    if (arr == NULL) {

        printf("malloc failed\n");
        return 1;
    }

    srand(time(NULL)); // Seed 
    generateRandomArray(arr, n, maxVal);


    clock_t start = clock(); 
    insertionSort(arr, n);
    clock_t end = clock();  


    double timeTaken = ((double)(end - start)) / CLOCKS_PER_SEC; 
    printf("Insertion Sort Time: %f seconds\n", timeTaken);

    free(arr); 
    return 0;
}
